test/gurka/test_recost.cc

```diff
@@ -0,0 +1,309 @@
+  };
+
+  // setup a callback for the recosting to tell us about the new label each made
+  bool called = false;
+  sif::LabelCallback label_cb = [&called](const sif::EdgeLabel& label) -> void { called = true; };
+
+  // build up the costing object
+  auto costing = sif::CostFactory().Create(Costing::auto_);
+
+  // those percentages are bonkers
+  EXPECT_THROW(sif::recost_forward(*reader, *costing, edge_cb, label_cb, -90, 476), std::logic_error);
+
+  // this edge id is valid but doesnt exist
+  EXPECT_THROW(sif::recost_forward(*reader, *costing, []() { return baldr::GraphId{123456789}; },
+                                   label_cb),
+               std::runtime_error);
+
+  // this edge id is not valid
+  sif::recost_forward(*reader, *costing, []() { return baldr::GraphId{}; }, label_cb);
+  EXPECT_EQ(called, false);
+
+  // this path isnt possible with a car because the second edge doesnt have auto access
+  called = false;
+  EXPECT_THROW(sif::recost_forward(*reader, *costing, edge_cb, label_cb), std::runtime_error);
+  EXPECT_EQ(called, true);
+
+  // go in the reverse direction
+  api = gurka::route(map, "F", "A", "pedestrian", {}, reader);
+  edge_itr = api.trip().routes(0).legs(0).node().begin();
+
+  // this path isnt possible with a car because the first node is a gate
+  called = false;
+  EXPECT_THROW(sif::recost_forward(*reader, *costing, edge_cb, label_cb), std::runtime_error);
+  EXPECT_EQ(called, true);
+
+  // travel only on pedestrian edges
+  api = gurka::route(map, "B", "D", "pedestrian", {}, reader);
+  edge_itr = api.trip().routes(0).legs(0).node().begin();
+
+  // it wont be able to evaluate any edges because they are all pedestrian only
+  called = false;
+  EXPECT_THROW(sif::recost_forward(*reader, *costing, edge_cb, label_cb), std::runtime_error);
+  EXPECT_EQ(called, false);
+}
+
+TEST(recosting, error_request) {
+  auto config = gurka::detail::build_config("foo_bar", {});
+  auto reader = std::make_shared<baldr::GraphReader>(config.get_child("mjolnir"));
+  valhalla::tyr::actor_t actor(config, *reader, true);
+
+  try {
+    actor.route(R"({"costing":"auto","locations":[],"recostings":[{"frosting":"chocolate"}]})");
+    FAIL() << "No costing should have thrown";
+  } catch (const valhalla_exception_t& e) { EXPECT_EQ(e.code, 127); }
+
+  try {
+    actor.route(R"({"costing":"auto","locations":[],"recostings":[{"costing":"foo"}]})");
+    FAIL() << "Wrong costing should have thrown";
+  } catch (const valhalla_exception_t& e) { EXPECT_EQ(e.code, 125); }
+
+  try {
+    actor.route(R"({"costing":"auto","locations":[],"recostings":[{"costing":"auto"}]})");
+    FAIL() << "No name should have thrown";
+  } catch (const valhalla_exception_t& e) { EXPECT_EQ(e.code, 127); }
+
+  try {
+    actor.route(R"({"costing":"auto","locations":[],"recostings":[{"name":"foo"}]})");
+    FAIL() << "No costing should have thrown";
+  } catch (const valhalla_exception_t& e) { EXPECT_EQ(e.code, 127); }
+}
+
+TEST(recosting, api) {
```

Comment:
this checks that the protobuf objects have the right recosting information set and that they general trend in numbers is correct (cars faster than bikes etc)